### PR TITLE
Feat: dealing with coinjoin blocked utxos

### LINKED ISSUE #8394: Do not register into round if too many utxos are banned
If a user has banned more than 40% of total non-private utxos or sum of banned utxos amounts is greater than 40% then they should not register to the next round for a certain period of time (1 hour).

It could work as a prevention of unknown bugs like https://github.com/trezor/trezor-suite/issues/8307

### PR DESCRIPTION
**Pull request by szymonlesisz**
## Description

Based on https://github.com/trezor/trezor-suite/pull/8416

Some of the utxos might be temporary banned (for max 6 hours)

If the ratio of banned utxos is higher than 40% of total utxos **or** sum of banned utxos amounts is greater than  40% of all amounts then account should not be registered in to round until the ratio becomes lower (depending on `sentenceEnd` in coinjoin prison)

New `SessionPhase` enum is added (BlockedUtxos) and handled in suite (translations)

## Related Issue

Resolve https://github.com/trezor/trezor-suite/issues/8394

## Screenshots:
image

### CODE REVIEW

**Comment by szymonlesisz** [comment]
rebased, resolved conflicts

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 1]
```diff
@@ -135,8 +135,46 @@ export const getAccountCandidates = ({
             return [];
         }
 
+        // filter out InputLongBanned utxos until we know how to deal with them...
+        const [_, whitelistedUtxos] = arrayPartition(
+            account.utxos,
+            utxo =>
+                prison.isDetained(utxo.outpoint)?.errorCode ===
+                WabiSabiProtocolErrorCode.InputLongBanned,
+        );
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/selectRound.ts:144]: why `arrayPartition` instead of basic filter?
```suggestion
        const whitelistedUtxos = account.utxos.filter(
  (utxo) =>
    prison.isDetained(utxo.outpoint)?.errorCode !==
    WabiSabiProtocolErrorCode.InputLongBanned
);
```

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 2]
```diff
@@ -198,7 +236,7 @@ export const getAccountCandidates = ({
             return groups;
         }, {} as Record<SessionPhase, Parameters<typeof setSessionPhase>[0]>);
 
-        Object.values(eventGroups).forEach(setSessionPhase);
+        Object.values(eventGroups).forEach(group => setSessionPhase(group));
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/selectRound.ts:239]: why this change?

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 3]
```diff
@@ -135,8 +135,46 @@ export const getAccountCandidates = ({
...
+        );
+
+        // collect known InputBanned
+        const bannedUtxos = whitelistedUtxos.filter(
+            utxo =>
+                prison.isDetained(utxo.outpoint)?.errorCode ===
+                WabiSabiProtocolErrorCode.InputBanned,
+        );
+
+        if (bannedUtxos.length > 0) {
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/selectRound.ts:153]: this condition does not apply in case of blame rounds, is that correct?

File: packages/coinjoin/src/enums.ts [hunk 4]
```diff
@@ -10,6 +10,7 @@ export enum SessionPhase {
     RetryingRoundPairing = 17,
     AffiliateServerOffline = 18,
     CriticalError = 19,
+    BlockedUtxos = 20,
```
Comment (matejkriz) [inline packages/coinjoin/src/enums.ts]: I'm afraid this can cause some troubles cc @dahaca  
https://github.com/trezor/trezor-suite/blob/feat/cj-blocked-utxos/packages/suite/src/utils/wallet/coinjoinUtils.ts/#L319

**Comment by matejkriz** [comment]
@szymonlesisz This is the related issue, right? https://github.com/trezor/trezor-suite/issues/8394 But @hynek-jina is suggesting 20 % instead of 40 % and 30 minutes detention instead of 1 hour? Can it be static or should it be configurable via message-system? Basically, it sounds like a static value to me, we should just determine it through testing.

**Comment by hynek-jina** [comment]
I don't have a strong opinion.. feel free to use values from @szymonlesisz

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 2 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts:239]: because foreach loop is using 3 parameters (item, index, array) and setSessionPhase is expecting only one.
i've added a test with check `toHaveBeenLastCalledWith` to check if my new message is passed correctly

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 3 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts:153]: right, inputs detained for blame rounds are using different code (not WabiSabiProtocolErrorCode.InputBanned)

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 1 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts:144]: see comment:
```
// filter out InputLongBanned utxos until we know how to deal with them...
```

currently they are stored in `_` until we know :)

**Comment by szymonlesisz** [comment]
>  But @hynek-jina is suggesting 20 % instead of 40 % and 30 minutes detention instead of 1 hour?

well, i remember we spoke about 40 :) we were also thinking about passing this value via messaging system (not implemented here)

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 1 above)
Comment (matejkriz) [reply packages/coinjoin/src/client/round/selectRound.ts:144]: I see :)

File: packages/coinjoin/src/enums.ts (hunk 4 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/enums.ts]: not sure, message is displayed properly from what i tested

but i guess this logic needs to be done better at some point

**Comment by szymonlesisz** [comment]
rebase on top of https://github.com/trezor/trezor-suite/pull/8582

**Comment by matejkriz** [comment]
> Rebasing failed, please rebase manually.

@szymonlesisz automatic rebase can't work if there are any conflicts

**Review by matejkriz — APPROVED** [review]
